### PR TITLE
Add support for apps to Singularity 3.0

### PR DESCRIPTION
**Pull request by bauerm97**
**Description of the Pull Request (PR):**

As title says, this PR adds apps to Singularity v3.0.0 in a very "plugin-like" manner. This will give us a nice clean way of building apps as a plugin when plugin support is fully introduced in 3.1, this is very much a prototype of that functionality (except linked statically instead of dynamically)

### CODE REVIEW

File: src/pkg/syplugin/build.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+// BuildPlugin is the interface for plugins on the build system
+type BuildPlugin interface {
+	Name() string
+	HandleSection(string, string)
+	HandleBundle(*types.Bundle)
+	HandlePost() string
+}
+
+// func init() {
```
Comment (yhcote) [inline src/pkg/syplugin/build.go:114]: Do you want to remove that ?

File: src/plugins/apps/apps.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+func appMeta(b *types.Bundle, a *App) string {
+	return filepath.Join(appBase(b, a), "/scif/")
+}
+
+func appData(b *types.Bundle, a *App) string {
+	return filepath.Join(b.Rootfs(), "/scif/data/", a.Name)
+}
+
+func writeFile(path string, perm os.FileMode, s string) error {
```
Comment (yhcote) [inline src/plugins/apps/apps.go:293]: what about ioutil/WriteFile()?

File: src/pkg/syplugin/build.go (hunk 1 above)
Comment (bauerm97) [reply src/pkg/syplugin/build.go:114]: Yea, I'll make a PR removing the comments once we tag beta

**Review by yhcote — APPROVED** [review]
Looks OK to me.

**Pull request merged**